packages/devtools_app/lib/src/flutter/controllers.dart

```diff
@@ -207,53 +137,17 @@ class _InheritedProvider extends InheritedWidget {
   const _InheritedProvider({@required this.data, @required Widget child})
       : super(child: child);
 
-  final ProvidedControllers data;
+  final ProvidedCommonControllers data;
 
   @override
   bool updateShouldNotify(_InheritedProvider oldWidget) =>
       oldWidget.data != data;
-
-  @override
-  InheritedElement createElement() => _DisposeAfterNotifyElement(this);
 }
 
-/// An [Element] that disposes its [_oldData] on the frame after after
-/// Notifying clients of a data change.
+/// Superclass for all DevTools controller providers.
 ///
-/// This allows clients to unregister listeners from the old data before it is
-/// disposed, and avoid exceptions caused by unregistering from disposed
-/// listeners.
-class _DisposeAfterNotifyElement extends InheritedElement {
-  _DisposeAfterNotifyElement(_InheritedProvider widget) : super(widget);
-
-  @override
-  _InheritedProvider get widget => super.widget;
-
-  ProvidedControllers _oldData;
-
-  @override
-  void updated(_InheritedProvider oldWidget) {
-    if (oldWidget.data != widget.data) {
-      _oldData = oldWidget.data;
-      super.updated(oldWidget);
-    }
-  }
-
-  @override
-  void notifyClients(_InheritedProvider oldWidget) {
-    super.notifyClients(oldWidget);
-    // For stateful widgets that depend on Controllers.of(context) to subscribe
-    // directly, we don't need a postframe callback and we can just  dispose the
-    // old data.
-    // For ValueListenableBuilder widgets, we need the postframe callback to
-    // wait until the builders have a chance to build and update their
-    // listeners.
-    SchedulerBinding.instance.addPostFrameCallback((_) {
-      _oldData?.dispose();
-    });
-  }
-}
-
-mixin OfflineControllerMixin<T> {
-  FutureOr<void> processOfflineData(T offlineData);
+/// For an example implementation, see `src/example/provided_controller.dart`.
+abstract class ControllerProvider extends StatefulWidget {
```

Comment:
Devtools is now depending on package:provider. I expect that will let you reduce the boilerplate defining all these providers manually.
https://pub.dev/packages/provider

Reply:
good call! eliminated a lot of code. PTAL.